pkg/col/coldata/bytes.go

```diff
@@ -17,161 +17,170 @@ import (
-	// data. Note that we assume that offsets are non-decreasing, and with every
-	// access of this Bytes we will try to maintain this assumption.
-	offsets []int32
-
-	// maxSetLength specifies the number of elements set by the user of this
-	// struct. This enables us to disallow unordered sets (which would require
-	// data movement). Also, this helps us maintain the assumption of
-	// non-decreasing offsets.
-	maxSetLength int
+// element describes a single []byte value. If a value doesn't exceed
+// BytesMaxInlineLength (30 bytes on 64 bit systems), then the whole value is
+// stored within the element; otherwise, the element "points" at a region in
+// Bytes.buffer where the non-inlined value is stored.
+//
+// If the value is inlined, then the layout of element is used as follows:
+//
+//            24-byte header | 6-byte padding
+//   element: .............................. | length | true
+//   Bytes.buffer: N/A
+//
+// where 30 dots describe the inlinable space followed by a single byte for the
+// length followed by a boolean 'true' indicating an inlined value.
+//
+// If the value is non-inlined, then the layout of element is used as follows:
+//
+//                                             padding
+//   element: .offset. | ..len... | ..cap... | xxxxxx | x | false
+//   Bytes.buffer: xxxxxxxx | offset .... | xxxxxxxx
+//
+// where first 24 bytes contain our custom "header" of a byte slice that is
+// backed by Bytes.buffer. The following 7 bytes (the padding and the
+// inlinedLength byte) are not used, and then we have a boolean 'false'
+// indicating a non-inlined value.
+type element struct {
+	header sliceHeader
+	_      [6]byte
+	// inlinedLength contains the length of the []byte value if it is inlined.
+	// Since the length of that value is at most BytesMaxInlineLength (30), it
+	// can definitely fit into a single byte.
+	inlinedLength byte
+	// inlined indicates whether the byte value is inlined in the element.
+	inlined bool
+}
 
-	// isWindow indicates whether this Bytes is a "window" into another Bytes.
-	// If it is, no modifications are allowed (all of them will panic).
-	isWindow bool
+// sliceHeader describes a non-inlined byte value that is backed by
+// Bytes.buffer.
+//
+// We do not just use []byte for two reasons:
+// 1. we can easily overwrite the whole struct when a value is inlined without
+// causing any issues with the GC (because our sliceHeader doesn't contain an
+// actual slice, GC doesn't have to track whether the backing array is live or
+// not);
+// 2. our slice header remains valid even when Bytes.buffer is reallocated. If
+// we were to use []byte, then whenever that reallocation occurred, we would
+// have to iterate over all non-inlined values and update the byte slices to
+// point to the new Bytes.buffer.
+type sliceHeader struct {
+	// bufferOffset stores the offset into Bytes.buffer where the corresponding
+	// bytes value begins.
+	bufferOffset int
+	len          int
+	cap          int
```

Comment:
Drive by comment, should we be explicit to use `int64` here? Since the size of `int` is platform dependent. On 32-bit platform, this would mean the `sliceHeader` would be 12 bytes instead of 24 bytes intended.